### PR TITLE
Split PyTensor instantiations into multiple files

### LINKED ISSUE #4741: Building PyTensor is slow
We currently instantiate 240 tensor types in `DataStructures/Tensor/Python/Tensor.cpp`, which takes quite a long time (couple of minutes, depending on the machine). Not sure how to speed that up though (suggestions welcome).

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Allows compiling them in parallel and reduces memory usage.

Fixes #4741 and (hopefully) issues on CI.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/Tensor/Python/InstantiateScalar.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/Tensor/Python/Tensor.tpp"
```
Comment (knelli2) [inline src/DataStructures/Tensor/Python/InstantiateScalar.cpp:4]: Do you also need the `.hpp` file here as well?

Comment (nilsvu) [reply src/DataStructures/Tensor/Python/InstantiateScalar.cpp:4]: It's included in the tpp 🤷‍♂️ this file just does instantiations so I think this is fine

Comment (nilsvu) [reply src/DataStructures/Tensor/Python/InstantiateScalar.cpp:4]: Oh this is the scalar file.. wasn't sure here either, but doesn't matter much I think

Comment (knelli2) [reply src/DataStructures/Tensor/Python/InstantiateScalar.cpp:4]: No it doesn't really matter. I thought we might have a guideline for this in our docs, but I couldn't find one. So this seems ok.

Comment (wthrowe) [reply src/DataStructures/Tensor/Python/InstantiateScalar.cpp:4]: I think when the tpp+hpp case first came up none of us could figure out what the right answer was, so we just left it out of the guidelines and let people do what they felt like.

**Review by nilsdeppe — APPROVED** [review]
LGTM

**Pull request merged**